compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2647,6 +2670,45 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         span.is_desugaring(DesugaringKind::QuestionMark)
             && self.tcx.is_diagnostic_item(sym::From, trait_def_id)
     }
+
+    /// Structurally compares two types, modulo any inference variables.
+    ///
+    /// Returns `true` if two types are equal, or if one type is an inference variable compatible
+    /// with the other type. A TyVar inference type is compatible with any type, and an IntVar or
+    /// FloatVar inference type are compatible with themselves or their concrete types (Int and
+    /// Float types, respectively). When comparing two ADTs, these rules apply recursively.
+    pub fn same_type_modulo_infer(&self, a: Ty<'tcx>, b: Ty<'tcx>) -> bool {
+        let (a, b) = self.resolve_vars_if_possible((a, b));
```

Comment:
Is this here so that we can return `false` for more cases? because otherwise the infer var case would just work.

Reply:
yes, if the infer var was constrained somewhere else since the last time it was resolved, we don't want to return true, or else we'll just suggest something wrong.

Reply:
Can you come up with a test that will ensure this?

Reply:
After removing this line, the following test regresses, erroneously suggesting to remove the semicolon.

```
fn baz(x: u64) -> u32 {
    x * 2;
}
```

I added this `resolve_vars_if_possible` to suppress erroneous suggestions.